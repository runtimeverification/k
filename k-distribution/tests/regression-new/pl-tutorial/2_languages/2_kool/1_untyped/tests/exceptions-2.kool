// This tests throwing and catching objects.  This is, in fact, how exceptions
// are expected to be handled in object-oriented languages.

class Exception {
  var v;
  method Exception(v) {
    print("Exception ", v, " thrown!\n");
    this.v = v;
  }
  method get() {
    return v;
  }
}

class Exception2 extends Exception {
  method Exception2(v) {
    Exception(v);
  }
}

class Main {

  method foo() {
    try {
      throw new Exception(5);
      print(17);                          // should not be printed
    } catch(e) {                          // "Exception2 e" in typed KOOL 
      throw new Exception2(e.get() + 2);  // unreachable in typed KOOl!
    }
    throw new Exception(-1);              // should not be reached
  }

  method Main() {
    try {
      foo();
    } catch(e) {
        print(e.get(),"\n");              // should print 7 (5 in typed KOOL)
    }
  }

}

// Exception 5 thrown!
// Exception 7 thrown!
// 7
