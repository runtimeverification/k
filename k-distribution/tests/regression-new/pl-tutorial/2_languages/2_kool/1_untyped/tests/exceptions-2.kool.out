Exception 5 thrown!
Exception 7 thrown!
7
